### PR TITLE
feat(BIND) - Impl all REST endpoints calls and the operations to bind/re-bind/unbind and update the name of the app

### PR DESCRIPTION
**Pull request by camilamacedo86**
## Motivation
* https://issues.jboss.org/browse/AEROGEAR-8986
* https://issues.jboss.org/browse/AEROGEAR-8916

## What
- Re-order the reconcile resources in db and app controller for them be created in the respective order
- Impl the REST endpoints call for bind/unbind/re-bind and update the name
- Add info in the readme
- Create a make command to run the project locally
- Operator config validation added

## Why
- Attend the requirements
- Avoid issues
- Make easier to do the local tests

## How
- Add new command in the Makefile
- Add rest endpoint calls 
- Changes in the controllers

## Verification Steps
1. Create a new namespace
2. Add 3 nodeJS app by the OCP catalogue with all required labels following an example.

### Bind configMap
1. Run the command make run-local 
2. Check that the operator was created
3. Check that the configMaps will be created with success

Screenshot 2019-04-12 at 10 24 36

### Bind App
After the above check see that the apps were created in the REST API by calling the endpoint GET http://mobile-security-service-app.<clusterIP>.nip.io/api/apps

Screenshot 2019-04-12 at 10 24 44

### Bind Delete app in rest service and configMap
1. Go to the OCP console and remove the mobilesecurityservice:bind label from DeploymentConfig and Pod YAML
2. After a few minutes check that configMap was removed and that app is no longer returned in the REST service. 

Screenshot 2019-04-12 at 10 28 55
 
Screenshot 2019-04-12 at 10 29 01

### Bind Update name of an app
1. Go to the OCP console and edit the value of the label app in DeploymentConfig and Pod YAML
2. Check in the deployments and pod that the label was changed
3.  Run the command make run-local again 
4. Check that the name was changed in the REST API after few minutes
Screenshot 2019-04-12 at 10 37 37
Screenshot 2019-04-12 at 10 34 11

### Operator config validation 
Remove the clusterHost IP in the step to configure it added in the CR and re-run again.

Screenshot 2019-04-12 at 12 00 04

## Checklist:

- [X] Code has been tested locally by PR requester
- [ ] Changes have been successfully verified by another team member 

## Progress

- [x] Finished task
- [ ] TODO

## Additional Notes

* We are not ensuring in this PR that the bind actions will be called when the changes are applied in the OCP and all tests are made calling the run-local again. We will track a task to do ensure that. 
* In future pr changes regards refectories will be required

### CODE REVIEW

File: README.adoc [hunk 1]
```diff
@@ -170,6 +171,54 @@ $ oc create rolebinding developer-mobile-security-service --role=mobile-security
...
+  # Watch filter setup for the Bind CR
+  # The watchKeyLabelSelector and watchValueLabelSelector defines the label key:value which will be watched by this type
+  # The watchNamespaceSelector defined an specific namespace for be watched by this type
+  # IMPORTANT: If no values be configured here the type will watch in the same namespace where the bind type was applied
+  # NOTES:
+  #     If just the key and label be filled then the bind will check in all namespaces the deployments with them
+  #     If just the watchNamespaceSelector be filled then the bind will consider all from this namespace only
+  #     Also, both can be filled and the filter will be aggregated, in this way the bind will watch the namespace and filter by the key:value selector as well
+  watchKeyLabelSelector: owner
+  watchValueLabelSelector: mdc
```
Comment (wei-lee) [inline README.adoc]: why do we need these for the bind operation? The operator should only need to watch the CRs?

File: README.adoc [hunk 2]
```diff
@@ -170,6 +171,54 @@ $ oc create rolebinding developer-mobile-security-service --role=mobile-security
...
+  watchNamespaceSelector:
+----
+
+===== How the Bind/Operator knows what application should be created/updated or delete in the REST Service API?
+
+The Bind/Operator will consider that all pod which are with the `key:value` defined in the following attributes in its  ./deploy/crds/mobile-security-service_v1alpha1_mobilesecurityservicebind_cr.yaml[CR] are running the applications which are bind to th Service.
```
Comment (wei-lee) [inline README.adoc]: why pods are involved here? For client apps, there are no pods running.

File: README.adoc [hunk 3]
```diff
@@ -170,6 +171,54 @@ $ oc create rolebinding developer-mobile-security-service --role=mobile-security
...
+
+The app name and id are the values of the respectively labels in the pods specified in the ./deploy/crds/mobile-security-service_v1alpha1_mobilesecurityservicebind_cr.yaml[MobileSecurityServiceBind CR]. Following these attributes.
+
+[source,yaml]
+----
+  # The appNameLabel spec defines the key label which will have the name of the app
+  appNameLabel: "app"
+  # The appIdLabel spec defines the key label which will have the id of the app
+  appIdLabel: "appId"
+----
```
Comment (wei-lee) [inline README.adoc]: I am not sure why this is necessary. The CR could have the appname/app id. Why does it need to query the pods? There are no pods running.

Comment (camilamacedo86) [reply README.adoc]: @wei-lee the master branch has the README updated. However, I think is good we chat by BJ.

Comment (camilamacedo86) [reply README.adoc]: How the MDC will create apps for the users whitout pods? How to bind a app to a service without  the app? Any app that be running in OCP will have a pod.

File: README.adoc (hunk 2 above)
Comment (camilamacedo86) [reply README.adoc]: I understand that is not the client app that we bind to the Service, but the cloud app which will have client apps managed by the Service.  The service is not bound to a client app at all. We are creating the apps which will have many client apps(versions). Am I not right?

File: README.adoc (hunk 1 above)
Comment (camilamacedo86) [reply README.adoc]: I schedule a time for we chat over it. But my understand is:
1. By the MDC the user will create an app in OCP
2. This app will be bind to services
3. The bind operator is responsible to check the apps created in the cluster which are bind to the Service and create/update/delete the app in the Service.

File: README.adoc (hunk 2 above)
Comment (wei-lee) [reply README.adoc]: no, that is not right. We will bind the client app to the service. Why do we need to bind the cloud app to the service?

Comment (camilamacedo86) [reply README.adoc]: It is fine after our chat. I am doing a new pr with it working as defined. I will ping you. Thank you.

**Pull request merged**